sys/dev/hwpmc/hwpmc_arm64_md.c

```diff
@@ -133,9 +140,22 @@ pmc_save_user_callchain(uintptr_t *cc, int maxsamples,
 
 	for (count = 1; count < maxsamples; count++) {
 		/* Use saved lr as pc. */
-		r = fp + sizeof(uintcap_t);
-		if (copyin((void * __capability)r, &pc, sizeof(pc)) != 0)
-			break;
+#ifdef COMPAT_FREEBSD64
+		if (SV_CURPROC_FLAG(SV_CHERI | SV_LP64) == SV_LP64) {
+			r = fp + sizeof(r64);
+			if (copyin(__USER_CAP(r, sizeof(r64)), &r64,
+			    sizeof(r64)) != 0)
+				break;
+			pc = r64;
+		} else
+#endif
+		{
+			r = fp + sizeof(uintcap_t);
+			if (copyin((void * __capability)r, &pc, sizeof(pc)) !=
```

Comment:
copyincap? Not that it hugely matters, the `&= ~1` will tag clear the sentry anyway (and we export it as a ptraddr_t to userspace), but should probably at least be consistent with fp for the copyin.

Reply:
I explicitly didn't do a copyincap (and even noted it in the commit log) as we don't need the tag so wanted to avoid copying in a valid capability I didn't need.

Reply:
Not sure I agree with that, it's a bit weird (and technically less efficient). If you're really set on that then it at least deserves a clear comment, not buried in a commit message, because as it stands the code looks wrong, or at least confusing due to the mismatch.

Reply:
It would perhaps be more clear if we changed the hwpmc api to push down ptraddr_t all the way to this level so it is clear we are returning addresses and not pointers.  A comment would not be bad though, and for now at least can clarify that we are only returning addresses anyway.

Reply:
If efficient matters though this should probably be using fuecap instead of copyin to avoid a pointless loop.  And that might be something to upstream (using fueword instead of copying)